schema.c

```diff
@@ -57,6 +57,8 @@ static __thread char tl_errmsg[BUFSIZ] = {0};
 
 #define READ_BUF_SIZE 512
 
+static sch_node *last_valid_schema = NULL;
```

Comment:
Not sure I like the addition of a global variable like this - is it actually used in more than one subroutine? If it needs to survive multiple calls to a single subroutine you can declare it static within that subroutine. But I would be worried whether multiple threads might be running for example using this same variable.

Reply:
I didn't realise that this routine is recursive. I think you could also make it an extra parameter on the recursive routine? Also, you should probably remove the earlier static definition as a global variable. To be honest, I don't know what making it atomic is actually achieving, that only helps in multi-thread situation when the value of the variable has meaning to both instances. In this case, it is surely something that needs to be independent in each thread (if there are indeed multiple threads possibly running this code)?